### ISSUE TITLE
CORS: preflight is needed if *not* all headers are simple

### ISSUE THREAD

**Issue by bluetech** [servo/servo#9472]
Unless I'm mistaken/misreading the code, according to https://www.w3.org/TR/cors/#cross-origin-request-with-preflight-0 the conditional in the following line should be negated:
https://github.com/servo/servo/blob/46b3eb653579a40632f91497a3d48f1d7fbd40cc/components/script/cors.rs#L80

**Comment by KiChjang** [comment]
There's also another possibility - we don't use W3C standards. Instead, we use WHATWG standards. Here's the link to the corresponding section https://fetch.spec.whatwg.org/#cors-preflight-fetch-0.

Also, if we are following W3C, then your observation is incorrect. The code you indicated is following this part of the algorithm:

> If the following conditions are true, follow the simple cross-origin request algorithm:
>    The request method is a simple method and the force preflight flag is unset.
>    Each of the author request headers is a simple header or author request headers is empty.

This is located in https://www.w3.org/TR/cors/#cross-origin-request.

**Comment by bluetech** [comment]
Thanks for pointing me at the correct spec - I just searched "cors spec" and hit the first one :) However the whatwg spec if is a bit too complicated for a passer-by like me, and I doubt the two specs say the exact opposite on such a detail, so I'll refer to the W3C one still.

I'm sorry to insist, but I still think the code is wrong, so I'll risk trying again. The code is:

```
if req.headers.iter().all(|h| is_simple_header(&h)) {
    req.preflight_flag = true;
}
```

Which I think means that if all headers are simple, a preflight request is needed. My understanding is that what it should say is: if there is at least one *non* simple header, *then* a preflight request is needed. And I think that's what the paragraph you quoted says ("Simple Cross-Origin Request" is opposite to "Cross-Origin Request with Preflight"). Or to quote another paragraph:

> Go to the next step if the following conditions are true:
>     - For request method there either is a method cache match or it is a simple method and the force preflight flag is unset.
>     - For every header of author request headers there either is a header cache match for the field name or it is a simple header.
> Otherwise, make a preflight request.

**Comment by KiChjang** [comment]
Ack, you're right. I've cross-referenced the situation with the WHATWG spec and it's located [here](https://fetch.spec.whatwg.org/#concept-main-fetch). Specifically,
> request's unsafe-request flag is set and either request's method is not a simple method or a header in request's header list is not a simple header

>    Set request's response tainting to "cors" and request's redirect mode to "error".

>    The result of performing an HTTP fetch using request with the CORS flag and CORS-preflight flag set. If the result is a network error, clear cache entries using request.

Both of the specs do say that if there's at least one header that is not a simple header, then a preflight fetch is required.

**Comment by Tycheon** [comment]
Hi there =)

I'd like to contribute. I've coded for quite awhile, but only ever on my own projects, so I've never done a pull request before.

If I'm following this properly, the change that needs to happen is that a preflight request flag needs to be set if any of the headers are not simple. Quickly checking the code shows that is_simple_header() just returns a bool, so we should be able to just change the line to:

if req.headers.iter().any(|h| !is_simple_header(&h)) {

The issues that I'm running into are:
1 - I'm not sure how to test to make sure it's actually doing what it should be, and
2 - I'm not sure how to actually do a pull request 

I don't suppose anyone would be able to point the completely clueless in the right direction?

**Comment by jdm** [comment]
It would make sense to run `./mach test-wpt tests/wpt/web-platform-tests/XMLHttpRequest` and `./mach test-wpt tests/wpt/web-platform-tests/cors`, as those should cover the change here.

**Comment by jdm** [comment]
As for the pull request, Github's got pretty good documentation: https://help.github.com/articles/creating-a-pull-request/ . Let us know if there's something more specific you're confused about!

**Comment by Tycheon** [comment]
Thank you so much for the quick response =)

I have to say, that I'm a little shocked by both the Rust and Servo communities; Everyone seems extremely patient, and just as friendly. It's also amazing that you have these E-easy issues, to ease people like me into this. It's so appreciated, you have no idea.

**Comment by Tycheon** [comment]
Hmmm... that one-line change causes the following to fail in the tests/wpt/web-platform-tests/cors suite:

  ▶ Unexpected subtest result in /cors/redirect-origin.htm:
  │ FAIL [expected PASS] remote (null) to local (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (none) to local (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (null) to remote (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (none) to remote (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (null) to remote2 (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (none) to remote2 (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ FAIL [expected PASS] remote (http://www1.web-platform.test:8000) to remote (*), expect to fail
  │   → assert_false: Got response expected false got true
  │ 
  │ shouldFail/</client.onreadystatechange<@http://web-platform.test:8000/cors/redirect-origin.htm:148:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  └ Test.prototype.step_func/<@http://web-platform.test:8000/resources/testharness.js:1405:1

I'm assuming that the tests are flawed (ie. that they reflect the originally flawed understanding of the spec), but am digging to find out for sure.

**Comment by jdm** [comment]
No, I would expect the tests to be correct in this case. They were written independent of Servo's implementation of the spec by other web browser developers.

**Comment by jdm** [comment]
It's possible that our implementation wasn't performing a CORS check previously and now it is, and it's also possible that our CORS check implementation is incorrect and causing a test failure now.

**Comment by KiChjang** [comment]
In which case, the test expectations should be updated, and we should file an additional issue in fixing it.

**Comment by Tycheon** [comment]
I've gone back and forth between the original line, and the new line. It seems to be that tests are failing due to getting response that they shouldn't.

I'm thinking that is_simple_header() in /components/script/cors.rs is to blame, but my knowledge of the rest of the code isn't quite sufficient to know for sure.

So here's the question, that function definition is: 

```
fn is_simple_header(h: &HeaderView) -> bool {
    // FIXME: use h.is::<HeaderType>() when AcceptLanguage and
    // ContentLanguage headers exist
    match &*h.name().to_ascii_lowercase() {
        "accept" | "accept-language" | "content-language" => true,
        "content-type" => match h.value() {
            Some(&ContentType(Mime(TopLevel::Text, SubLevel::Plain, _))) |
            Some(&ContentType(Mime(TopLevel::Application, SubLevel::WwwFormUrlEncoded, _))) |
            Some(&ContentType(Mime(TopLevel::Multipart, SubLevel::FormData, _))) => true,
            _ => false,
        },
        _ => false,
    }
}
```


The problem (as I see it, which might be horrifically flawed) is that we want to set the flag if *any* of the headers are not simple. However, in checking for whether or not a header is simple, we return true from is_simple_header if the header matches a certain type... and false for everything else.... including **invalid headers**?

**Comment by bluetech** [comment]
The `is_simple_header` function is following the spec: https://fetch.spec.whatwg.org/#simple-header

I don't think there are really "invalid headers" (at this point) - perhaps you mean custom/non-standard headers? In which case, yes, if you add such to the request, they are not simple and a preflight request is needed.

Here is a more high level simplified description than the raw specs, in case you haven't read it yet, which explains the rationale behind the "simple header" rule: http://www.html5rocks.com/en/tutorials/cors/#toc-adding-cors-support-to-the-server

**Comment by bluetech** [comment]
For even higher level rationale, see http://tools.ietf.org/html/rfc6454, especially [3.4.2](http://tools.ietf.org/html/rfc6454#section-3.4.2).

Also note, that if you do a simple cross-origin XHR `GET`, without doing any `setRequestHeader`s, then the `header-list` at the time of the discussed check should be empty + the simple headers appended in the algorithm: https://fetch.spec.whatwg.org/#concept-fetch. So, if you try that, which headers do you see failing the test? (I am too lazy to try myself).

**Comment by Tycheon** [comment]
Haven't had a whole lot of time to work on this, and what time I did spend was spent trying to wrap my head around flow of execution. 

**I'm not entirely certain how to perform that simple cross-origin XHR Get. Is that from within the js test suite? or from within the servo codebase proper?**

Also, I was able to pin down exactly which tests were failing within redirect-origin.html ... the bizarre thing is that if the only change I make to the current code in cors.rs is to go from:

```
if req.headers.iter().all(|h| is_simple_header(&h)) {
    req.preflight_flag = true;
}
```

to

```
if req.headers.iter().any(|h| !is_simple_header(&h)) {
    req.preflight_flag = true;
}
```

Then the tests fail. Which immediately made me think that the flag is now being set for some of those tests, when before it wasn't. So I further changed it to:

```
if req.headers.iter().any(|h| !is_simple_header(&h)) {
    panic!();
    req.preflight_flag = true;
}
```

Just to quickly see if the whole thing would grind to a halt at those failing tests. It would be a quick and dirty way to check to see if that flag is suddenly being set for those tests, and somehow causing them to fail.

It isn't. 

The panic happens for other tests, but not for any of the redirect-origin.htm tests.... and even for the other tests, a panic only happens for a couple. Conversely, swapping back to the original code, and injecting a panic each time the flag is set:

```
if req.headers.iter().all(|h| is_simple_header(&h)) {
    panic!();
    req.preflight_flag = true;
}
```

causes a panic on the overwhelming majority of the tests. I'm not certain which ones, within each .htm test suite, as I'm not 100% certain how to hook into the servo code from within the js testing suite...

At any rate. Not entirely certain exactly where the problem is, but I'm still digging around.

**Comment by jdm** [comment]
APIs that rely on the CORS  implementation are easiest to observe when invoked from JS, rather than from Servo proper. The simplest example is an XMLHttpRequest on one origin performing a GET request to another origin with a custom request header present (eg. `X-foo`). http://mxr.mozilla.org/servo/source/tests/wpt/web-platform-tests/cors/request-headers.htm?force=1 is a decent instance of this in the testsuite.

**Comment by jdm** [comment]
So all of the cases that report problems with the change are cases that were expected to return no response, ie. a CORS violation was expected. However, it appears that the violation is only being reported when the preflight is requested (ie. `req.preflight_flag = true`), so now the code believes that the request is fine, hence the test failures. The solution will not be to ensure that there is a preflight for these cases, but to ensure that there are still appropriate CORS checks that occur outside of the preflight response.

**Comment by jdm** [comment]
I think the integration point you're looking for is in XMLHttpRequest::Send (`components/script/dom/xmlhttprequest.rs`), where we start making decisions about CORS things. This corresponds to JS calling the `send` method on an `XMLHttpRequest` object.

**Comment by Tycheon** [comment]
Thanks @jdm =)

I've shoots booked all weekend, so I'll dig deeper on Monday or Tuesday.

I figured that a simple one-liner would be an easy place to start, to pop my "first PR cherry". Guess nothing is ever as simple as it appears. Thanks for pointing me in the right direction in the codebase!

**Comment by Tycheon** [comment]
So, I'm looking through XMLHttpRequest::Send() and the associated spec at https://xhr.spec.whatwg.org/#the-send()-method ... both code and spec seem to line up nicely until step 9. Is there a CORS portion of the Send() spec that's located somewhere else? Or is it all in https://fetch.spec.whatwg.org/#cors-preflight-fetch-0 ?

Also, with regards to the preflight_flag, it is *not* being set for the tests that are failing. I'm guessing that it is being set in the current master branch, but there's a lot of noise, so it's hard to tell.

I figured that since this original issue was a result of a minor misinterpretation of the spec, that the reason for tests failing once it's fixed, is that there's a further misinterpretation of the spec somewhere down the line... I'm just not certain where to look.

**Comment by Ms2ger** [comment]
The xhr spec used to specify all this explicitly, before the fetch spec existed; now it defers to the fetch spec, but our code still reflects the previous structure.

**Comment by jdm** [comment]
XMLHttpRequest::check_cors seems to be the place where we would expect the relevant checks to occur. Unfortunately, CORSRequest::http_fetch_async (and therefore CORSRequest::http_fetch) seems to do no checking at all if no preflight is requested.

**Comment by jdm** [comment]
There's also allow_cross_origin_request which is called from XMLHttpRequest::process_headers_available, which might be relevant here.

**Comment by Tycheon** [comment]
@jdm Thanks so much for the message in IRC!! Adding --log-mach - to the end of the command for running the specific test:

./mach test-wpt tests/wpt/web-platform-tests/cors/redirect-origin.htm --log-mach -

Works like a charm for printing out debug information. The 7 misbehaving tests within redirect-origin.htm were definitely having the preflight_flag set to true previously, and are now no longer having those same flags set to true.

Now that I can actually peek inside structs in fns, I'll hopefully be able to figure out why that is. =)

**Comment by Tycheon** [comment]
I'm at the point where I could use another nudge, if possible. I tested to ensure that the preflight_flag *was* being set to true for the 7 failing tests, but, with the logic change as part of this issue, it's no longer being set.

I've been trying to look for differences between the execution path when preflight_flag = true vs false, and the most glaring is that in xmlhttpresponse.rs, the function XMLHttpRequest::process_headers_available is called regardless of whether or not that preflight_flag is set.

However, the match block is behaving differently for each:

 ```
   fn process_headers_available(&self, cors_request: Option<CORSRequest>,
                                 gen_id: GenerationId, metadata: Metadata) -> Result<(), Error> {

        if let Some(ref req) = ors_request {
            match metadata.headers {
                Some(ref h) if allow_cross_origin_request(req, h) => {},
                _ => {
                    self.process_partial_response(XHRProgress::Errored(gen_id, Error::Network));
                    return Err(Error::Network);
                }
            }
        }

        // XXXManishearth Clear cache entries in case of a network error
        self.process_partial_response(XHRProgress::HeadersReceived(gen_id,
                                                                   metadata.headers,
                                                                   metadata.status));
        Ok(())
    }
```

when preflight_flag is set to false (as it is with the noted change in cors.rs) then metadata.headers contains actual headers, and the second process_partial_response() is fired (for these tests, AccessControlAllowOrigin is triggering Any. I had initially thought that the associated comment in allow_cross_origin_request() in cors.rs might have something to do with that, but it looks like that's not something that is implemented yet?)

However, when preflight_flag is set to true (as it is in master right now), then metadata.headers is None. So the first process_partial_response() is fired, returning an error, which is what the test suite is expecting, so all tests pass.

My problem is that I'm having difficulty tracking where those response headers are being set... any pointers?

**Comment by jdm** [comment]
Yes, those come from the `load` function in `components/net/http_loader.rs`.

**Comment by Tycheon** [comment]
Here's where preflight_flag = true, vs preflight_flag = false diverge in that function (right afterwards, the metadata.headers are None for pf = true):

```
if response.status().class() == StatusClass::Redirection {
            if let Some(&Location(ref new_url)) = response.headers().get::<Location>() {
                // CORS (https://fetch.spec.whatwg.org/#http-fetch, status section, point 9, 10)
                if let Some(ref c) = load_data.ors {
                    if c.preflight {
                        return Err(
                            LoadError::Cors(
                                url,
                                "Preflight fetch inconsistent with main fetch".to_owned()));
                    } else {
                        // XXXManishearth There are some CORS-related steps here,
                        // but they don't seem necessary until credentials are implemented
                    }
                }
```

When the preflight_flag is true, an Err is returned due to an inconsistency. However, if the preflight_flag is false, it runs into the else block, loops back through the entire function again, and provides the test harness with a response that it shouldn't

**Comment by KiChjang** [comment]
Fixed in #9667.
